### PR TITLE
Generate box proposals op

### PR DESCRIPTION
**Pull request by samikama**
This PR adds GenerateBoxProposals op for accelerating FPN based networks such as MaskRCNN. It requires #28745 to be merged.

### CODE REVIEW

**Comment by samikama** [comment]
Pinging @tfboyd @tatianashp. This is the third part of ops to accelerate FPN based networks.

**Comment by aaroey** [comment]
Hi @alextp and @pengchongjin, would you help to review the API part of this PR? Thanks

File: tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt
Comment (alextp) [inline tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt]: Use the in_arg and out_arg fields to populate inputs and outputs: https://github.com/tensorflow/tensorflow/blob/273981699d92c1aa08ca52af99ee187ce5aaa5e1/tensorflow/core/framework/api_def.proto#L79

File: tensorflow/core/ops/image_ops.cc
Comment (alextp) [inline tensorflow/core/ops/image_ops.cc]: Why are these attributes and not inputs?

File: tensorflow/tools/api/golden/v1/tensorflow.pbtxt
Comment (alextp) [inline tensorflow/tools/api/golden/v1/tensorflow.pbtxt]: Probably belongs in tf.image, not tf

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt
Comment (samikama) [reply tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt]: This file is auto-generated by tensorflow tools. Do you want me to change them? Wouldn't it break next time they are autogenerated again?

File: tensorflow/core/ops/image_ops.cc
Comment (samikama) [reply tensorflow/core/ops/image_ops.cc]: Because they don't depend on data? They are model parameters and fixed at the model construction time.

File: tensorflow/tools/api/golden/v1/tensorflow.pbtxt
Comment (samikama) [reply tensorflow/tools/api/golden/v1/tensorflow.pbtxt]: Implementation is in tf.image. This is another auto-generated file. What kind of changes should I do so that api generator puts it tf.image?

**Comment by aaroey** [comment]
@samikama I think the interface of NonMaxSuppressionV2 is fixed, then why does this PR depend on #28745?

**Comment by aaroey** [comment]
@pengchongjin, would you help to provide review for the interface from the perspective of object detection?

Thanks.

**Comment by aaroey** [comment]
@smit-hinsu, would you please help to look at the implementation after API review is done? Thanks.

**Comment by aaroey** [comment]
@wujingyue just in case you're interested in this op.

File: tensorflow/core/ops/image_ops.cc
Comment (pengchongjin) [inline tensorflow/core/ops/image_ops.cc]: Nit: '.' in the end of the sentence.

File: tensorflow/core/ops/image_ops.cc
Comment (pengchongjin) [inline tensorflow/core/ops/image_ops.cc]: Nit: '.' in the end of the sentence.

File: tensorflow/core/ops/image_ops.cc
Comment (pengchongjin) [inline tensorflow/core/ops/image_ops.cc]: Any particular reason to make Num_Achors be the second dimension?

File: tensorflow/core/ops/image_ops.cc
Comment (pengchongjin) [inline tensorflow/core/ops/image_ops.cc]: Are these sorted? Can you comment on it?

File: tensorflow/core/ops/image_ops.cc
Comment (pengchongjin) [inline tensorflow/core/ops/image_ops.cc]: Any comment on what does this do?

File: tensorflow/core/ops/image_ops.cc
Comment (samikama) [reply tensorflow/core/ops/image_ops.cc]: It is not 2d. There are 4 times num_anchors elements in the array. Each 4 points representing an anchor box

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (aaroey) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: The pattern:
```
context->allocate_temp(...)
SetZero<<<...>>>(...)
```
occurs >10 times, can we add a helper method for that?

**Review by aaroey — CHANGES_REQUESTED** [review]

**Comment by samikama** [comment]
Sorry for not responding. Vacation and a few other things got in the way. I will get back to this as soon as possible.

File: tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt
Comment (samikama) [reply tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt]: Done

File: tensorflow/core/ops/image_ops.cc
Comment (samikama) [reply tensorflow/core/ops/image_ops.cc]: I converted these to tensors as requested. But considering the op, it will just cause an extra overhead since it will require many constant ops to be scheduled unnecessarily.

File: tensorflow/tools/api/golden/v1/tensorflow.pbtxt
Comment (samikama) [reply tensorflow/tools/api/golden/v1/tensorflow.pbtxt]: It should be done now.

File: tensorflow/core/ops/image_ops.cc
Comment (samikama) [reply tensorflow/core/ops/image_ops.cc]: Docs are moved to api_def files.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

File: tensorflow/python/ops/image_ops_impl.py
Comment (aaroey) [inline tensorflow/python/ops/image_ops_impl.py]: @alextp could you help to review this change?

Comment (alextp) [reply tensorflow/python/ops/image_ops_impl.py]: I'm a little swamped; can you find someone else? Or I can look at it on friday.

**Review by aaroey — CHANGES_REQUESTED** [review]

File: tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt
Comment (samikama) [reply tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt]: This affects the output shape. I was told before that the attributes that change the output shape should be attributes not inputs. If I do it, this will be a data dependent op. Do you really want that?

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: it is in the system headers. float.h

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: It was to prevent stream synchronization needed with nms calls, however current event processing infrastructure of TF is insufficient for doing such kind of asynchronous processing thus idea was abandoned. I reverted it back to sync op.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (aaroey) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Use camel case for function names, also use more descriptive name e.g. ApplyOffset.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (aaroey) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Fix the reference to `KA`.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (aaroey) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Applying minus sign twice makes it harder to read, maybe add an `SubtractOffset` func for it.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (aaroey) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Please comment the shape of d_img_info_vec and d_scores.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (aaroey) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: We already have `dev_boxes_keep_flags` to tell which boxes to keep, why do we need to pass `dev_sorted_scores` and set scores of all not-selected boxes to min(float)?

Also from the code below we'll use `cub::DeviceSelect::Flagged` and `dev_boxes_keep_flags` to filter out those scores, so setting the score to min(float) becomes not useful?

**Review by aaroey — CHANGES_REQUESTED** [review]
Thanks for the fix, just a few more minor comments.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Hi @aaroey , that is correct. I removed score reference.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: They are described in op documentation but I added it here as well.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: Done

**Review by aaroey — DISMISSED** [review]

**Comment by ppwwyyxx** [comment]
`correct_transform_coords` is not necessarily correct. The transform logic in this inference op is correct only when it is the inverse of the transform logic during training. 
The history of the "correct_transform_coords", I believe, comes from a [bug in the py-faster-rcnn project](https://github.com/rbgirshick/py-faster-rcnn/blob/781a917b378dbfdedb45b6a56189a31982da1b43/lib/fast_rcnn/bbox_transform.py#L30-L61) (and also many derived projects from it), where the box encoding and decoding are not the exact inverse of each other.
In other words, `correct_transform_coords=True` is correct only when the encoding during training uses the same implementation in [py-faster-rcnn](https://github.com/rbgirshick/py-faster-rcnn/blob/781a917b378dbfdedb45b6a56189a31982da1b43/lib/fast_rcnn/bbox_transform.py#L10-L24), but different implementations do exist. For example, in Facebook we're [dropping](https://github.com/pytorch/pytorch/blob/master/caffe2/operators/generate_proposals_op_util_boxes.h#L43-L81) the old implementation in py-faster-rcnn/Detectron with the `legacy_plus_one` option.

Also I don't know why `correct_transform_coords=False` could be useful in any situation except to match results of legacy buggy implementation. I couldn't understand why an op added to TensorFlow should take care of certain buggy 3rd-party implementation of papers.

**Comment by aaroey** [comment]
@samikama I feel the same way, if `correct_transform_coords` is added only for a legacy but buggy usage I don't think we should add that.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Review by pengchongjin — CHANGES_REQUESTED** [review]
Interfaces look good to me. Could you please comment on whether input/output scores are sorted?

**Comment by samikama** [comment]
@ppwwyyxx and @aaroey Unfortunately, probably due to rush to publication and recognition, authors are publishing buggy code and that code is becoming a reference in the community which sometimes results in new code to be compared against. I believe that buggy code @ppwwyyxx referencing is still quite prevalent in the community and still used same way. I added that flag to be able to satisfy requirements of users still keeping that offset. I will shortly remove it.
@ppwwyyxx is there an errata or public page that shows that offset was a bug that never should have been there in the first place so that if users complain about op behaving differently then we can point them to and ask fixing their models?

**Comment by ppwwyyxx** [comment]
The original repo (py-faster-rcnn) is unmaintained. There is only one report I found about this bug  https://github.com/rbgirshick/py-faster-rcnn/issues/852.

It was acknowledged as a bug in https://github.com/facebookresearch/Detectron/issues/298#issuecomment-373942633 (3).

I'm not aware of other resources that mention this problem. But we plan to explain and fix these historical issues together with the next version of Detectron to be released later this year.

**Comment by samikama** [comment]
@ppwwyyxx So this bug is not widely known and not really easy to find out. That is in line with my understanding of the situation. @aaroey @alextp are you sure you want to remove compatibility flag, considering that most networks inspired by detectron will likely to have the bug still and will not be able to use this op without changing their networks?

**Comment by samikama** [comment]
@tatianashp as discussed in meeting last week, I removed the support for the legacy implementation. If it turned out to be a problem, last commit can be reverted and other changes can be reapplied.
@pengchongjin I added relevant comments to the api_def files since I have been asked to move the documentation to that files above. Inputs scores can be unsorted, op will sort them internally, output will be sorted by scores in descending order.
To be merged after #30893

**Comment by samikama** [comment]
@pengchongjin @alextp @aaroey As requested, I removed the legacy mode option. I still believe it is a bad choice since the **bug** is not documented, not fixed, not known by anybody other than a handful of people and still being used in models. I believe removing it will lead to confusion and head scratching. 
Do you have any other concerns?

**Review by aaroey — DISMISSED** [review]

**Review by aaroey — DISMISSED** [review]

**Comment by samikama** [comment]
@aaroey, Just pushed the fixes for api files. Thanks

**Comment by samikama** [comment]
@aaroey @rthadur please merge this after #30893 since it contains fixes that are used by this op as well.

File: tensorflow/tools/api/golden/v1/tensorflow.pbtxt
Comment (rthadur) [inline tensorflow/tools/api/golden/v1/tensorflow.pbtxt]: Internal review comments : 'The op needs a visibility:HIDDEN to avoid polluting the toplevel TF namespace'

**Review by rthadur — CHANGES_REQUESTED** [review]

**Review by pkulzc — DISMISSED** [review]
LGTM.

**Comment by rthadur** [comment]
@pkulzc  can we pull this in ?

**Comment by samikama** [comment]
@rthadur please wait until I make another commit since during the review process, there were some api changes in consumers and at current state it would require unnecessary operations to wire it up.  I will reformat the data layouts to match users.

File: tensorflow/tools/api/golden/v1/tensorflow.pbtxt
Comment (aaroey) [reply tensorflow/tools/api/golden/v1/tensorflow.pbtxt]: Please fix this @samikama

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by samikama** [comment]
Hi, Yes this is still being worked on. Waiting for some internal updates.

**Comment by samikama** [comment]
@aaroey  @pkulzc This should be good to go now.

File: tensorflow/tools/api/golden/v1/tensorflow.pbtxt
Comment (samikama) [reply tensorflow/tools/api/golden/v1/tensorflow.pbtxt]: Done

**Review by aaroey — APPROVED** [review]

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (aaroey) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: @samikama remove all CHECKs and return proper error instead. It's not allowed and breaking the merge tests.

**Review by aaroey — CHANGES_REQUESTED** [review]
@samikama remove all CHECKs and return proper error instead. It's not allowed and breaking the merge tests.

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: @aaroey Is this a new requirement? I am seeing still there are a lot of references to CHECKs in TF code base. Can you be a bit more specific on which tests it is breaking?

Comment (aaroey) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: This is not new. It's required that new code (since >1 years ago I believed) should avoid checks. The merge process is blocked because of this check.

Comment (samikama) [reply tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: That is strange since I can see commits as recent as last month that has same CHECK_* ops in the trunk. Are you sure that it is blocking the merge? If so how are some of the links below passed these tests? Pretty much all of them created after this pr is opened. Perhaps there should be some PR requirements document that everybody should adhere since if we use style/code similar to what is in TF or recent PRs, they are being marked as blockers. Could either somebody fix these in the TF so that other contributors can't use them as reference or can there be some requirements document created that everybody follows that document and prevent confusion and ease up non-google contributors life and improve their enthusiasm.

Thanks,
 
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/rocm/rocm_gpu_executor.cc#L421
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/host/host_gpu_executor.cc#L45
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/stream_executor/gpu/redzone_allocator.cc#L275
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/python/lib/core/py_func.cc#L193
https://github.com/tensorflow/tensorflow/blob/6c32a22396b7ac9f9b9d77b147113281b689b345/tensorflow/python/eager/pywrap_tensor.cc#L121

**Comment by alextp** [comment]
Can this op live in tf-addons?

**Comment by samikama** [comment]
Hi, @alextp I don't intend to move this to tf-addons, especially not after more than 6 months of going back on forth in this PR.

**Review by aaroey — DISMISSED** [review]

**Review by aaroey — DISMISSED** [review]

File: tensorflow/core/kernels/generate_box_proposals_op.cu.cc
Comment (rthadur) [inline tensorflow/core/kernels/generate_box_proposals_op.cu.cc]: `third_party/tensorflow/core/kernels/generate_box_proposals_op.cu.cc:398:9: error: no matching function for call to 'max'
        max(cub_sort_temp_storage_bytes, cub_select_temp_storage_bytes);`

can you please check above error

**Review by rthadur — CHANGES_REQUESTED** [review]

**Review by aaroey — DISMISSED** [review]

**Review by aaroey — DISMISSED** [review]

**Review by aaroey — APPROVED** [review]

File: tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt
Comment (ppwwyyxx) [inline tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt:20]: why there are 5 elements each row if it only has height, width and scale? This makes no sense

Comment (wbudd) [reply tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt:20]: Yeah, [generate_box_proposals_op.cu.cc#L115~118](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/image/generate_box_proposals_op.cu.cc#L115) contains the following...
```C
const float img_height = d_img_info_vec[5 * image_index + 0];
const float img_width = d_img_info_vec[5 * image_index + 1];
const float min_size_scaled = min_size * d_img_info_vec[5 * image_index + 2];
```
...at which point one would expect `5 * image_index + 3` to be assigned to some `foo` and a `5 * image_index + 4` to some `bar`, but this never happens. These two float elements effectively ending up being nothing more than some weird unused padding, it seems.

I suspect they were used as extra attributes at some earlier development stage, and code wasn't properly updated upon their removal?

@aaroey @pengchongjin 
Any reason why this didn't seem like cause for concern during (pre-merge) review?

Comment (wbudd) [reply tensorflow/core/api_def/base_api/api_def_GenerateBoundingBoxProposals.pbtxt:20]: Actually, I guess the padding might be intended as a bank conflict avoidance tactic. Either way, implementation details shouldn't distort the public API; and if they do anyway there should be some comment/documentation mentioning the underlying reasoning.

**Pull request merged**
